Sources/SpeziHealthKit/HealthKit.swift

```diff
@@ -68,129 +69,215 @@ import SwiftUI
-            HKHealthStore.isHealthDataAvailable(),
-            """
-            HealthKit is not available on this device.
-            Check if HealthKit is available e.g., using `HKHealthStore.isHealthDataAvailable()`:
-            
-            if HKHealthStore.isHealthDataAvailable() {
-                HealthKitHealthStore()
-            }
-            """
-        )
-
-        self.healthStore = HKHealthStore()
+    
+    /// Creates a new instance of the ``HealthKit`` module, with an empty configuration.
+    public convenience init() {
+        self.init { /* intentionally empty config */ }
     }
 
-    static func didAskForAuthorization(for sampleType: HKSampleType) -> Bool {
-        // `alreadyRequestedSampleTypes` is always just written using `healthKitSampleTypesIdentifiers`, so this can stay
-        // non-isolated as UserDefaults is generally thread-safe.
-        UserDefaults.standard.alreadyRequestedSampleTypes.contains(sampleType.identifier)
-    }
 
+    /// Configures the HealthKit module.
     public func configure() {
-        for healthKitDataSourceDescription in initialHealthKitDataSourceDescriptions {
-            execute(healthKitDataSourceDescription)
+        Task {
+            for component in exchange(&pendingConfiguration, with: []) {
+                await component.configure(for: self, on: self.standard)
+            }
         }
     }
+    
+    
+    // MARK: HealthKit authorization handling
 
-
-    /// Displays the user interface to ask for authorization for all HealthKit data defined by the ``HealthKitDataSourceDescription``s.
+    /// Requests authorization for accessing all HealthKit data types defined in the ``HealthKit-swift.class`` module's current data access requirements list.
+    /// Once the initial configuration of the ``HealthKit-swift.class`` module has completed, this list will consist of all data types defined by the individual configuration components,
+    /// i.e., for example anything required by a ``CollectSample`` component, or explicitly requested via ``RequestReadAccess`` or ``RequestWriteAccess``.
+    ///
+    /// Calling this function will also activate all data collecting components from the module configuration (e.g., ``CollectSample``)
+    /// which require read access to data types for which the user hadn't already be prompted in the past.
+    /// If all necessary authorizations have already been requested (regardless of whether the user granted or denied access),
+    /// this function will not present any UI to the user, but will still activate the configuration components.
     ///
-    /// Call this function when you want to start HealthKit data collection.
+    /// - Note: There is no need for an app itself to keep track of whether it already requested health data access; the ``HealthKit-swift.class`` takes care of this.
+    ///
+    /// - Important: You need to call this function at some point during your app's lifecycle, ideally soon after the app was launched.
```

Comment:
Do I understand this correctly that this always has to be called on every app launch instead of just once for the onboarding of a user and all subsequent launches start to collect the data automatically?

Reply:
Yes.

In a sense, this was already a requirement before, but it was never stated explicitly.
(Imagine e.g. adding a new `CollectSample` to the config, with a sample type you haven't requested read access for yet, or in general needing read access to some new sample type somewhere in the app. You would've needed to call `askForAuthorization()` outside your onboarding to trigger the HealthKit aurhorization screen in that scenario, so putting the call only into the onboarding would only have worked if you never wanted to change anything.)

IMO, this new approach (of always requiring the call) is cleaner overall, since the app doesn't need to keep track of (or somehow determine) whether the user was already asked, which is already handled by the HealthKit module.

Reply:
Though actually now that I think a little bit more about it, it would of course in theory be possible to auto-start data collection for all non-manual, already-authorized data sources. Though this could lead to a situation where an app forgets that it needs to also call `askForAuthorization()` outside its initial onboarding, and then will never receive data for any new/changed config queries.